### ISSUE TITLE
nestedScrollEnabled prop not working in ScrollView (Android) with horizontal direction

### ISSUE THREAD

**Issue by minhthong095** [facebook/react-native#21436]
- [X] Review the documentation: https://facebook.github.io/react-native
  - [X] Search for existing issues: https://github.com/facebook/react-native/issues
  - [X] Use the latest React Native release: https://github.com/facebook/react-native/releases

## Environment
Binaries:
    npm: 5.6.0 - C:\Program Files\nodejs\npm.CMD
    IDEs:
      Android Studio: Version  3.1.0.0 AI-173.4907809
Package:
    "react": "16.5.0",
    "react-native": "0.57.1",
Device Info:
    Android Version: 5.0.2 LRX22G

## Description
I think _nestedScrollEnabled_ prop for _ScrollView_ only work in vertical. I don't know what's wrong in my code or something else. Just change little bit in height and width length and add/remove _nestedScrollEnabled_. 
Could anyone help me in this issue.

## Reproducible Demo
**VERTICAL**
```
<ScrollView >
        <View style={{ backgroundColor: 'red', borderRightColor: 'white', borderWidth: 3, width: 500, height: 500 }}>
          <View style={{ borderWidth: 2, borderColor: 'blue', backgroundColor: 'yellow', width: 100, height: 300 }}>
            <ScrollView  nestedScrollEnabled>
              <View style={{ borderWidth: 3, borderColor: 'black', backgroundColor: 'purple', width: 100, height: 100 }} />
              <View style={{ borderWidth: 3, borderColor: 'black', backgroundColor: 'purple', width: 100, height: 100 }} />
              <View style={{ borderWidth: 3, borderColor: 'black', backgroundColor: 'purple', width: 100, height: 100 }} />
              <View style={{ borderWidth: 3, borderColor: 'black', backgroundColor: 'purple', width: 100, height: 100 }} />
              <View style={{ borderWidth: 3, borderColor: 'black', backgroundColor: 'purple', width: 100, height: 100 }} />
            </ScrollView>
          </View>
        </View>
        <View style={{ backgroundColor: 'red', borderRightColor: 'white', borderWidth: 3, width: 500, height: 500 }}>
          <View style={{ borderWidth: 2, borderColor: 'blue', backgroundColor: 'yellow', width: 100, height: 300 }}>
            <ScrollView  nestedScrollEnabled>
              <View style={{ borderWidth: 3, borderColor: 'black', backgroundColor: 'purple', width: 100, height: 100 }} />
              <View style={{ borderWidth: 3, borderColor: 'black', backgroundColor: 'purple', width: 100, height: 100 }} />
              <View style={{ borderWidth: 3, borderColor: 'black', backgroundColor: 'purple', width: 100, height: 100 }} />
              <View style={{ borderWidth: 3, borderColor: 'black', backgroundColor: 'purple', width: 100, height: 100 }} />
              <View style={{ borderWidth: 3, borderColor: 'black', backgroundColor: 'purple', width: 100, height: 100 }} />
            </ScrollView>
          </View>
        </View>
      </ScrollView>
```
![vertical-scroll](https://user-images.githubusercontent.com/15071968/46280375-f2da6c80-c595-11e8-9808-13035b6252ae.gif)

**HORIZONTAL**
```
<ScrollView horizontal>
  <View style={{backgroundColor: 'red',borderRightColor: 'white', borderWidth: 3, width: 500, height: 500}}>
    <View style={{ borderWidth: 2, borderColor: 'blue', backgroundColor: 'yellow', width: 300, height: 100 }}>
      <ScrollView horizontal nestedScrollEnabled>
        <View style={{ borderWidth: 3, borderColor: 'black', backgroundColor: 'purple', width: 100, height: 100 }} />
        <View style={{ borderWidth: 3, borderColor: 'black', backgroundColor: 'purple', width: 100, height: 100 }} />
        <View style={{ borderWidth: 3, borderColor: 'black', backgroundColor: 'purple', width: 100, height: 100 }} />
        <View style={{ borderWidth: 3, borderColor: 'black', backgroundColor: 'purple', width: 100, height: 100 }} />
        <View style={{ borderWidth: 3, borderColor: 'black', backgroundColor: 'purple', width: 100, height: 100 }} />
      </ScrollView>
    </View>
  </View>
  <View style={{backgroundColor: 'red',borderRightColor: 'white', borderWidth: 3, width: 500, height: 500}}>
    <View style={{ borderWidth: 2, borderColor: 'blue', backgroundColor: 'yellow', width: 300, height: 100 }}>
      <ScrollView horizontal nestedScrollEnabled>
        <View style={{ borderWidth: 3, borderColor: 'black', backgroundColor: 'purple', width: 100, height: 100 }} />
        <View style={{ borderWidth: 3, borderColor: 'black', backgroundColor: 'purple', width: 100, height: 100 }} />
        <View style={{ borderWidth: 3, borderColor: 'black', backgroundColor: 'purple', width: 100, height: 100 }} />
        <View style={{ borderWidth: 3, borderColor: 'black', backgroundColor: 'purple', width: 100, height: 100 }} />
        <View style={{ borderWidth: 3, borderColor: 'black', backgroundColor: 'purple', width: 100, height: 100 }} />
      </ScrollView>
    </View>
  </View>
</ScrollView>
```
![horizontal-scroll](https://user-images.githubusercontent.com/15071968/46280353-e524e700-c595-11e8-992c-6a50d55cc224.gif)

**Comment by minhthong095** [comment]
Anyone? I believe it's a bug.

**Comment by esbenvb** [comment]
I have the same problem.

Code here https://snack.expo.io/@esbenvb/scrollview-tests

Android behavior https://vimeo.com/322265966

iOS behavior https://vimeo.com/322266040

```
  React Native Environment Info:
    System:
      OS: macOS 10.14.3
      CPU: (12) x64 Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
      Memory: 331.47 MB / 16.00 GB
      Shell: 3.2.57 - /bin/bash
    Binaries:
      Node: 10.12.0 - /usr/local/bin/node
      Yarn: 1.10.1 - /usr/local/bin/yarn
      npm: 6.4.1 - /usr/local/bin/npm
      Watchman: 4.9.0 - /usr/local/bin/watchman
    SDKs:
      iOS SDK:
        Platforms: iOS 12.1, macOS 10.14, tvOS 12.1, watchOS 5.1
      Android SDK:
        API Levels: 23, 25, 26, 27, 28
        Build Tools: 26.0.2, 26.0.3, 27.0.3, 28.0.2, 28.0.3
        System Images: android-25 | Google Play Intel x86 Atom, android-26 | Google Play Intel x86 Atom, android-28 | Google APIs Intel x86 Atom, android-28 | Google Play Intel x86 Atom
    IDEs:
      Android Studio: 3.3 AI-182.5107.16.33.5199772
      Xcode: 10.1/10B61 - /usr/bin/xcodebuild
    npmPackages:
      react: 16.6.3 => 16.6.3 
      react-native: 0.58.6 => 0.58.6 
```

**Comment by keenubee** [comment]
Confirm, I have the same issue with nested horizontal scrolling.

**Comment by aanah0** [comment]
Have the same issue

**Comment by joaom182** [comment]
Same issue here

**Comment by ThakurKarthik** [comment]
+1 any fix regarding this?

**Comment by smm76** [comment]
Having the same issue. nestedScrollEnabled property works fine for vertical/vertical or vertical/horizontal (and vice versa) but seems to have no effect for nested horizontal scrollviews inside of another horizontal scrollview. 
However, i found some kind of workaround for this. This is a bit hacky and does not work perfectly, as you have to actually drag the inner view, but it works. 
You can capture the onTouchStart, onTouchEnd and onTouchCancel Events of the inner scrollview and set scrollEnabled on the outer scrollview to true or false correspondingly.

**Comment by superKalo** [comment]
I'm experiencing the same issue. @smm76 yep, tried this hackish workaround too, but it doesn't feel 'fluid' enough.

Please reopen the issue.

**Comment by superKalo** [comment]
@zhongwuzw , can you please re-open this issue or please tell any other collaborator to re-open it, since the @react-native-bot closed it.

**Comment by superKalo** [comment]
For the brave hearts who fight with this issue, here's my workaround.

I have a horizontal `<FlatList />` (parent) and a horizontal `<ScrollView />` (child).

On my inner `ScrollView` (child) I am disabling the `<FlatList />` (parent) scroll `onTouchStart` and I am enabling it back `onScrollEndDrag`, similarly like @smm76 suggested. But additionally, I'm doing this throttle magic, which kind of solves all the little edge cases when the user doesn't immediately start dragging.

```javascript
import throttle from 'lodash/throttle';

...

<ScrollView
  horizontal={true}
  scrollEventThrottle={16}
  // ... bla bla

  onTouchStart={() => {
    if (Platform.OS === 'android') {
      this.props.onTouchStartDisableScrollOnAndroid(); // disables parent scroll

      // If the `onScrollEndDrag` event is not fired almost immediately
      // after this one (and user doesn't start dragging)
      // fire up `onTouchEndEnableScrollOnAndroid` and enable back
      // the parent horizontal view scroll.
      // This may happen when the user quickly taps on the component
      // but doesn't start dragging.
      this.throttledOnScrollEndDragOnAndroid = throttle(
          this.props.onTouchEndEnableScrollOnAndroid,  // enables parent scroll
          300,
          {
              leading: false,
          },
      );
      this.throttledOnScrollEndDragOnAndroid();
    }
  }}
  onScrollEndDrag={() => {
    if (Platform.OS === 'android') {
      this.props.onTouchEndEnableScrollOnAndroid(); // enables parent scroll

      if (this.throttledOnScrollEndDragOnAndroid) {
          this.throttledOnScrollEndDragOnAndroid.cancel();
      }
    }
  }}
>
  /* content */
</ScrollView>
```

**Comment by timorss** [comment]
on all the scrollviews add prop nestedScrollEnabled,
including the nested scrollview!

**Comment by fjmorant** [comment]
Same problem for me, with:

```
<FlatList horizontal nestedScrollEnabled={true}>
    <ScrollView horizontal nestedScrollEnabled={true}>
    </ScrollView>
</FlatList>
```

And when you try to scroll the child horizontally it scrolls the parent :(

**Comment by brunolemos** [comment]
Oh No. I'm facing this as well. Nested horizontal scrollviews doesn't work on Android. `nestedScrollEnabled` has no effect in this case. Still happening on react-native 0.61.

I made this Snack with a reproduction: https://snack.expo.io/@brunolemos/nested-scrollview

<div>
<img alt="gif-ios-works" width="320" src="https://user-images.githubusercontent.com/619186/66369405-ac6ed900-e972-11e9-9fac-899a5bb0e197.gif" />
<img alt="gif-android-doesnt-work" width="320" src="https://user-images.githubusercontent.com/619186/66369406-ac6ed900-e972-11e9-8512-30d6a9ee6eda.gif" />
</div>

<details>

<summary>See full code</summary>

```tsx
import React from 'react'
import { ScrollView, ScrollViewProps, View, ViewProps } from 'react-native'

function App() {
  return <Columns />
}

export default App

function Story(props: ViewProps) {
  return (
    <View
      {...props}
      style={[
        {
          width: 30,
          height: 30,
          borderRadius: 30 / 2,
          backgroundColor: 'blue',
        },
        props.style,
      ]}
    />
  )
}

function Stories(props: ScrollViewProps) {
  return (
    <ScrollView
      horizontal
      nestedScrollEnabled
      {...props}
      style={[
        {
          width: '100%',
          maxHeight: 30 + 2 * 2,
          backgroundColor: 'lightyellow',
        },
        props.style,
      ]}
    >
      <Story style={{ margin: 2 }} />
      <Story style={{ margin: 2 }} />
      <Story style={{ margin: 2 }} />
      <Story style={{ margin: 2 }} />
      <Story style={{ margin: 2 }} />
      <Story style={{ margin: 2 }} />
      <Story style={{ margin: 2 }} />
      <Story style={{ margin: 2 }} />
      <Story style={{ margin: 2 }} />
      <Story style={{ margin: 2 }} />
    </ScrollView>
  )
}

function Column(props: ViewProps) {
  return (
    <View
      {...props}
      style={[
        {
          paddingTop: 40,
          width: 200,
          height: '100%',
          backgroundColor: 'lightgreen',
        },
        props.style,
      ]}
    >
      <Stories />
    </View>
  )
}

function Columns(props: ScrollViewProps) {
  return (
    <ScrollView
      horizontal
      nestedScrollEnabled
      {...props}
      style={[
        { width: '100%', height: '100%', backgroundColor: 'lightblue' },
        props.style,
      ]}
    >
      <Column style={{ margin: 10 }} />
      <Column style={{ margin: 10 }} />
      <Column style={{ margin: 10 }} />
      <Column style={{ margin: 10 }} />
      <Column style={{ margin: 10 }} />
      <Column style={{ margin: 10 }} />
      <Column style={{ margin: 10 }} />
      <Column style={{ margin: 10 }} />
      <Column style={{ margin: 10 }} />
      <Column style={{ margin: 10 }} />
    </ScrollView>
  )
}
```

</details>

**Comment by brunolemos** [comment]
@andreicoman11 @makovkastar @aamalric @astreet I checked the file blame and noticed you contributed a lot to the [ReactHorizontalScrollView.java](https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/views/scroll/ReactHorizontalScrollView.java) file. I was wondering (hoping) if you have any insight of what could be causing the issue above and what the fix could look like?

**Comment by aamalric** [comment]
@brunolemos My contribution was just about preventing scrolling with D-Pad key events when scroll is disabled, nothing to do with nested scroll with touch events...
Looking at the code I don't really see why it's not working, there's no difference between vertical and horizontal scroll managers code for Android, they both call:

`ViewCompat.setNestedScrollingEnabled(view, value);`

https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/views/scroll/ReactHorizontalScrollViewManager.java#L157

https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/views/scroll/ReactScrollViewManager.java#L167

**Comment by ThakurKarthik** [comment]
I used React Native Swiper for this purpose.(Nested Scrolling)

**Comment by brunolemos** [comment]
### Workaround with nested `FlatList`

I investigated a little bit and I found out that `FlatList` has implemented some workarounds for nested case. If you have nested `FlatList` instead of nested `ScrollView`, you can do this:

`<FlatList {...props} renderScrollComponent={p => <ScrollView {...p} />} />`

If you don't override the scroll component like above, `FlatList` will render a simple `View` instead:

https://github.com/facebook/react-native/blob/d34bc5fa64a54dfc2e780461ee2997a4b17f8c65/Libraries/Lists/VirtualizedList.js#L1155-L1158

### Nested `ScrollView` is still unsolved

If you have nested `ScrollView` and not nested `FlatList`, that's harder to make a workaround. You can try the `onTouchStart` thing mentioned above or you can check `FlatList` source code to see how they did it. I believe it's this part:

https://github.com/facebook/react-native/blob/d34bc5fa64a54dfc2e780461ee2997a4b17f8c65/Libraries/Lists/VirtualizedList.js#L1444-L1450

_When the parent scrolls, they are calling some internal methods of the nested lists._

Seems to be complicated to implement yourself, so maybe try using `FlatList` if possible.

**Comment by brunolemos** [comment]
I found an easy workaround!

If we use the `ScrollView` from `react-native-gesture-handler` everything works as expected.

```tsx
import { ScrollView as GestureHandlerScrollView } from 'react-native-gesture-handler'

<GestureHandlerScrollView horizontal>
    <GestureHandlerScrollView horizontal />
</GestureHandlerScrollView>

// or 

<ScrollView horizontal>
    <GestureHandlerScrollView horizontal />
</ScrollVIew>
```

You can have nested horizontal `FlatList`s like this:

```tsx
<FlatList {...props} renderScrollComponent={p => <GestureHandlerScrollView {...p} />} />
```

Updated the snack with this workaround: https://snack.expo.io/@brunolemos/nested-scrollview

**Comment by smm76** [comment]
@brunolemos nice one! Just tried on a real device and it worked!
Weird anyway, how does the gesture-handler implementation differ from native ScrollView? 
Is this down to touch recognition?

**Comment by cklinx** [comment]
@brunolemos Tanks so much!

**Comment by brunolemos** [comment]
@cklinx do you still need this prop? also confirm that you are using the latest version of both `react-native` (0.61.2) and `react-native-gesture-handler` (1.4.1).

**Comment by cklinx** [comment]
@brunolemos the scroll is working but when the child's scrollview reach end the parent's scroll will not fire. You can see it in your expo example also so i don't think its! 

![horizontal scroll](https://user-images.githubusercontent.com/32547415/66546244-e2f84f80-eb3c-11e9-86f4-da037c9e2e69.gif)

**Comment by brunolemos** [comment]
.

**Comment by ertandrn** [comment]
I have the same problem. If you find the solution, please share.

**Comment by smm76** [comment]
@ertandrn try to use ScrollView Component of `react-native-gesture-handler` instead of the one from `react-native` as suggested by @brunolemos . This seems to be the only non-hacky solution atm and works as expected

**Comment by ertandrn** [comment]
@smm76 i tried, but the problem persists.

**Comment by smm76** [comment]
@ertandrn could you provide some code or a demo then?
To be sure, take a look at this expo snack: https://snack.expo.io/@brunolemos/nested-scrollview
This works for me.

**Comment by ertandrn** [comment]
@smm76  i tried the same code on the link but it doesn't work. (https://snack.expo.io/@brunolemos/nested-scrollview)

**Comment by berkaey** [comment]
same here

**Comment by berkaey** [comment]
after some tests, I found that if you give both internal views(cards for ex.) and scrollview content style pixel based width, it will work but if you give % width it just won't scroll but renders without an issue.

**Comment by mistikk** [comment]
@brunolemos your solution is working on snack but it is not working on emulator or real devices

**Comment by arshidkv12** [comment]
Same here

**Comment by willymulyana** [comment]
I stumble in similar issue. iOS is fine but when test on Android, horizontal `ScrollView` won't scroll. The horizontal `ScrollView` that set `overflow='hidden'` seem not working too.

My setup: horizontal `ScrollView` is inside vertical `ScrollView`, I have item list inside horizontal `ScrollView`. After sometime I manage to fix it, so my code is like following:
```
<ScrollView style={{flex:1}}>
   ...
   <ScrollView horizontal style={{width:W-40, overflow:'hidden', marginHorizontal:20}} >
      ...
   </ScrollView>
   ...
</ScrollView>
```
Some to point out for horizontal `ScrollView`:

- was using `flex:1` and on Android it will make the `ScrollView` will have content/children width, making so that it can't be swiped and `overflow:'hidden'`  not working

- with the use of horizontal margin, it can't use `width:'100%'` too as in Android and iOS will adopt parent's width, and while the swipe is working, it make the `overflow:'hidden'` seem not working

- not using nestedScrollEnabled prop

My environment:
"react": "16.9.0",
"react-native": "0.61.5",

**Comment by jjhampton** [comment]
The [workaround](https://github.com/facebook/react-native/issues/21436#issuecomment-539415308) provided by @brunolemos works for me on Android . Does anyone know why the ScrollView from `react-native-gesture-handler` works, and the core RN ScrollView does not?

**Comment by Kiesco08** [comment]
Same here, [this](https://github.com/facebook/react-native/issues/21436#issuecomment-539415308) worked for me.

**Comment by ThienMD** [comment]
> Same here, [this](https://github.com/facebook/react-native/issues/21436#issuecomment-539415308) worked for me.

It work, thanks

**Comment by safaiyeh** [comment]
> I found an easy workaround!
> 
> If we use the `ScrollView` from `react-native-gesture-handler` everything works as expected.
> 
> ```tsx
> import { ScrollView } from 'react-native'
> import { ScrollView as GestureHandlerScrollView } from 'react-native-gesture-handler'
> 
> <ScrollView horizontal>
>     <GestureHandlerScrollView horizontal />
> </ScrollVIew>
> ```
> 
> You can have nested horizontal `FlatList`s like this:
> 
> ```tsx
> <FlatList {...props} renderScrollComponent={p => <GestureHandlerScrollView {...p} />} />
> ```
> 
> Updated the snack with this workaround: https://snack.expo.io/@brunolemos/nested-scrollview
> 
> <img alt="gif-android-working" width="320" src="https://user-images.githubusercontent.com/619186/66381996-b9022a00-e990-11e9-901d-e8f3362ee648.gif">
> 
> Maybe someone can check what `react-native-gesture-handle` is doing and bring the same workaround to the core of `react-native`.

Appreciate the solution! This issue is a bit dated, if anyone still has this issue and none of these solutions work, feel free to open a new issue.

**Comment by scousino** [comment]
I mean shouldn't this issue be fixed in react-native core? The solution can't be just use a ScrollView from a different library.

**Comment by safaiyeh** [comment]
@scousino relooking at this 100% agree. Thanks for flagging. It should be fixed in core.

**Comment by Adigezalov** [comment]
I have a similar problem, but my structure is like this:
`<ScrollView horizontal><FlatList horizontal /></ScrollView>`
And the FlatList does not.
How can this problem be solved now?

**Comment by fjmorant** [comment]
Definitely I think this should be solved in the core

**Comment by kAVArd** [comment]
Need to fix nested horizontal scroll on Android in the core. `ScrollView` from `react-native-gesture-handler` works perfect!

**Comment by vikks** [comment]
For the ScrollViews to work as expected, nested or otherwise, you need to specify their boundaries. Boundaries are set in 
"**style**" property of the ScrollViews.

```javascript
import { ScrollView } from 'react-native'

//
<ScrollView 
     style={{ 
           width: "equal-to-or-less-than viewport width", 
           height: "equal-to-or-less-than viewport height"
     }}
    contentContainerStyle={{"..."}}
    horizontal
>
    <ScrollView 
           style={{
              width: "lower-than parent scrollview width",
              height: "lower-than parent scrollview height"
           }} 
           contentContainerStyle={{"..."}}
    >
    </ScrollView>
</ScrollVIew>
```

> NOTE: Always Set explicit boundaries(width/height) for Scrollview. So basically all sides of the border have to appear on the screen

**Comment by fabriziobertoglio1987** [comment]
https://github.com/facebook/react-native/pull/28138

**Comment by fabriziobertoglio1987** [comment]
pr https://github.com/facebook/react-native/pull/28138 would manually implement this feature, was closed in favour of pr https://github.com/facebook/react-native/pull/18299 which adds a prop for [ViewCompat.setNestedScrollingEnabled](https://developer.android.com/reference/androidx/core/view/ViewCompat#setNestedScrollingEnabled(android.view.View,%20boolean))

>**Enable or disable nested scrolling for this view.**

>If this property is set to true the view will be permitted to initiate nested scrolling operations with a compatible parent view in the current hierarchy. **If this view does not implement nested scrolling this will have no effect.** Disabling nested scrolling while a nested scroll is in progress has the effect of stopping the nested scroll.

By my experience writing pr https://github.com/facebook/react-native/pull/29025, the `onTouchEvent` in the `ReactScrollView` child does not include logic to handle nested scrolling


**<details><summary>CLICK TO OPEN TESTS RESULTS</summary>**
<p>

| **IOS** | **ANDROID** | 
|:-------------------------:|:-------------------------:|
| <img src="https://user-images.githubusercontent.com/619186/66369405-ac6ed900-e972-11e9-9fac-899a5bb0e197.gif" width="300" height="" /> | <img src="https://user-images.githubusercontent.com/619186/66369406-ac6ed900-e972-11e9-8512-30d6a9ee6eda.gif"  width="300" height="" /> |
</p>
</details>

The current ReactScrollView `onTouchEvent` only detects when user finger stops scrolling and then delegates to the parent view. 

https://github.com/facebook/react-native/blob/d79212120b7168015d3d0225ef372ed851a230fa/ReactAndroid/src/main/java/com/facebook/react/views/scroll/ReactScrollView.java#L315-L336

The solutions:

**1 - Enhacing the `onTouchEvent`**
using a logic similar to my pr https://github.com/facebook/react-native/pull/29025 which adds possibility to scroll TextInput that are included in a ScrollView

The logic detects: 
* the swipe direction (`horizontal` or `vertical`)
* delegates the swipe gesture to the parent component once the nested scroll view reaches the end


**<details><summary>CLICK TO OPEN TESTS RESULTS</summary>**
<p>

| **HORIZONTAL** | **AFTER** | 
|:-------------------------:|:-------------------------:|
| <img src="https://user-images.githubusercontent.com/24992535/86393696-b45b6f80-bc9d-11ea-97a6-f6fb7cdf92ca.gif" width="400" height="" /> | <img src="https://user-images.githubusercontent.com/24992535/86393701-b6253300-bc9d-11ea-96ae-f1d99da29c38.gif"  width="400" height="" /> |

</p>
</details>

https://github.com/facebook/react-native/blob/7b4f13455b1ec1c86746a3050dbaea3f47b57da6/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.java#L190-L211

**2 - Inherit this functionality from other classes (for ex. NestedScrollView)**
- inherit from classes like [NestedScrollView](https://developer.android.com/reference/androidx/core/widget/NestedScrollView) that include this functionalities

I'll troubleshoot and try to prepare a PR.

**Comment by fabriziobertoglio1987** [comment]
I would like to ask @tuncaulubilge that wrote pr https://github.com/facebook/react-native/pull/18299 which adds a prop [`nestedScrollEnabled`](https://developer.android.com/reference/androidx/core/view/ViewCompat#setNestedScrollingEnabled(android.view.View,%20boolean)) if he could give me feedback before starting to work on this, as it is a big change, but seems essential to use the functionality included in his pr https://github.com/facebook/react-native/pull/18299 :pray: 

>**Enable or disable nested scrolling for this view.**

>If this property is set to true the view will be permitted to initiate nested scrolling operations with a compatible parent view in the current hierarchy. **If this view does not implement nested scrolling this will have no effect.** 

seems to me that the problem here is missing implementation of the [`NestedScroll`](https://developer.android.com/reference/androidx/core/view/NestedScrollingChild#dispatchNestedScroll(int,%20int,%20int,%20int,%20int[])) functionality in the `ReactScrollView`

>This interface should be implemented by View subclasses that wish to support dispatching nested scrolling operations to a cooperating parent ViewGroup.

as my understanding when using `<ScrollView nestedScrollEnabled />` the prop will create a `final instance` of `NestedScrollingChildHelper`

```java
public class ReactScrollView extends ScrollView implements NestedScrollingChild {
    private NestedScrollingChildHelper helper = new NestedScrollingChildHelper(this);
}
```

>Classes implementing this interface should create a final instance of a NestedScrollingChildHelper as a field and delegate any View methods to the [NestedScrollingChildHelper](https://developer.android.com/reference/androidx/core/view/NestedScrollingChildHelper) methods of the same signature.
>For each View method that has a matching method signature in this class, delegate the operation to the helper instance in an overridden method implementation. This implements the standard framework policy for nested scrolling.

```java
public class ReactScrollView extends ScrollView implements NestedScrollingChild {
    private NestedScrollingChildHelper helper = new NestedScrollingChildHelper(this);
    
    @Override
    protected void onScrollChanged(int l, int t, int oldl, int oldt) {
        // our implementations of onScrollChanged
        super.onScrollChanged(l, t, oldl, oldt);
    }

    @Override
    public boolean isNestedScrollingEnabled() {
        return helper.isNestedScrollingEnabled();
    }

    @Override
    public void setNestedScrollingEnabled(boolean enabled) {
        helper.setNestedScrollingEnabled(enabled);
    }

    @Override
    public boolean startNestedScroll(int axes) {
        return helper.startNestedScroll(axes);
    }

    @Override
    public void stopNestedScroll() {
        helper.stopNestedScroll();
    }

    @Override
    public boolean hasNestedScrollingParent() {
        return helper.hasNestedScrollingParent();
    }

    @Override
    public boolean dispatchNestedScroll(int dxConsumed, int dyConsumed, int dxUnconsumed, int dyUnconsumed, int[] offsetInWindow) {
        return helper.dispatchNestedScroll(dxConsumed, dyConsumed, dxUnconsumed, dyUnconsumed, offsetInWindow);
    }

    @Override
    public boolean dispatchNestedPreScroll(int dx, int dy, int[] consumed, int[] offsetInWindow) {
        return helper.dispatchNestedPreScroll(dx, dy, consumed, offsetInWindow);
    }

    @Override
    public boolean dispatchNestedFling(float velocityX, float velocityY, boolean consumed) {
        return helper.dispatchNestedFling(velocityX, velocityY, consumed);
    }

    @Override
    public boolean dispatchNestedPreFling(float velocityX, float velocityY) {
        return helper.dispatchNestedPreFling(velocityX, velocityY);
    }
}
```

Other Links: 

https://stackoverflow.com/questions/35510250/how-to-properly-implement-nestedscrollingchild-on-a-webview
https://stackoverflow.com/questions/29885067/how-to-implement-nestedscrolling-on-android
https://stackoverflow.com/questions/30089084/android-issue-with-implementing-flings-in-nestedscrollview

**Comment by fabriziobertoglio1987** [comment]
Thanks a lot @isnotgood for pr https://github.com/facebook/react-native/pull/28138 which seems to implement this solution. 

I did not study much the ReactNativeGestureHandler `ScrollView horizontal` component https://github.com/facebook/react-native/issues/21436#issuecomment-539415308 which does not have this issue

I'm curious in the reason @isnotgood closed pr https://github.com/facebook/react-native/pull/28138, maybe performance limitation compared to ReactNativeGestureHandler?

>the lookup code could be performance bottleneck in some extreme cases and I'm willing to optimize it further if it will be considered

Tomorrow I will give a further look in the ReactNativeGestureHandler implementation to see if I can understand and apply the same solution to HorizontalScrollView.java

I am also considering creating a pull request to add information about this bug in the [ScrollView docs](https://reactnative.dev/docs/next/scrollview) so that the solution is easier to find. Thanks

**Comment by isnotgood** [comment]
@fabriziobertoglio1987 there was no problem, I just closed it as there was no interest from maintainers 😄 Feel free to resurrect the PR on your own if you think it's worth the effort.

**Comment by dincozdemir** [comment]
I re-opened the PR with the permission of @isnotgood

**Comment by scousino** [comment]
Just bumping in here to show people are still waiting for this fix. Can we get some maintainers to review this PR and get the fix in already.

**Comment by HoangTran0410** [comment]
I have tried to use ScrollView from `react-native-gesture-handler` and add `disableIntervalMomentum` props, then it worked!

```jsx
import { ScrollView } from 'react-native-gesture-handler'

<ScrollView horizontal disableScrollViewPanResponder>
    <ScrollView horizontal />
</ScrollVIew>
```

**Comment by faerylay** [comment]
i found this problem when we use nested scrollView to prevent horizontal or vertical scroll error but its doesnt work in android .so i make it like this (sry for my english )

<scrollView vertical>
<scrollView horizontal >  <--- this scrollview component is not using just prevent scrollview same props error (so just add scrollEnabled to false ) thats it no need to add     nestedScrollEnabled
<scrollView vertical>

</scrollView vertical>
</scrollView horizontal>
</scrollView vertical>

**Comment by russmenum** [comment]
Is there ever going to be a fix for this 2023 and still a bug

**Comment by vikashkeshri14** [comment]
here is the solution just add extra scrollView with different oriantation it will solve the issue ex
<ScrollView vertical>
  <ScrollView horizontal>
<ScrollView vertical></ScrollView>
</ScrollView>
</ScrollView>

**Comment by russmenum** [comment]
> here is the solution just add extra scrollView with different oriantation it will solve the issue ex `<ScrollView vertical> <ScrollView horizontal> <ScrollView vertical></ScrollView> </ScrollView> </ScrollView>`

Even that was not working, it seems there is a bug with respect to the Android gesture/swipe. If you are on a screen where an Android user can swipe back, the nested scroll layer is not accessible for most of the screen. While disabling swipe allows the nested scroll to work, ideally, there would be a way to fix them layering on top of each other incorrectly.

It is not like Android is going to abandon nav gestures and users expect them at this point

**Comment by sfsnaruto** [comment]
Small workaround which just works with scrollView
`import React, { Component } from 'react';
import { View, ScrollView } from 'react-native';

export default class App extends Component {
  constructor() {
		super();
		this.state = {
			enabled:true,
      enable2:true
		};
	}
   enableDisable = (en1, en2) =>{
    this.setState({
      	enabled:en1,
      enable2:en2
    })
    setTimeout(()=>{
      this.setState({
      	enabled:!en1,
      enable2:!en2
    })
    },500)
  }
  render() {
    return (
    	<ScrollView 
      horizontal
      nestedScrollEnabled
      scrollEnabled={this.state.enabled}
      persistentScrollbar
      onEndReach={()=>{

      }}
       onMomentumScrollEnd={()=>{
         console.log('renabled---')
         this.enableDisable(false, false)
         }}
     style={{width:300, height:300, flexDirection:'row'}}
						>
            <View style={{flexDirection:'row'}}>
						<View style={{width:60,backgroundColor:'violet'}}></View>
            	<View style={{width:602,backgroundColor:'violet'}}></View>
							<View style={{ width: 200, backgroundColor: 'red' }} >
								<ScrollView
                nestedScrollEnabled
								 horizontal
                  onMomentumScrollEnd={this.state.enabled2}
                  onScrollEndDrag={()=>{ this.enableDisable(false, false)}}
									style={{ margin: 10, height: 200, width:200, flexDirection:'row' }} >
									<View style={{ width: 200, backgroundColor: 'blue' }} />
									<View style={{ width: 200, backgroundColor: 'pink' }} />
									<View style={{ width: 200, backgroundColor: 'blue' }} />
									<View style={{ width: 200, backgroundColor: 'pink' }} />
									<View style={{ width: 200, backgroundColor: 'blue' }} />
									<View style={{ width: 200, backgroundColor: 'pink' }} />
									<View style={{ width: 200, backgroundColor: 'blue' }} />
									<View style={{ width: 200, backgroundColor: 'pink' }} />
									<View style={{ width: 200, backgroundColor: 'blue' }} />
									<View style={{ width: 200, backgroundColor: 'pink' }} />
									<View style={{ width: 200, backgroundColor: 'blue' }} />
									<View style={{ width: 200, backgroundColor: 'pink' }} />
								</ScrollView>
							</View>
               </View>
						</ScrollView>
           
    );
  }
}`

**Comment by Glaysonvisgueira** [comment]
> For the ScrollViews to work as expected, nested or otherwise, you need to specify their boundaries. Boundaries are set in "**style**" property of the ScrollViews.
> 
> ```js
> import { ScrollView } from 'react-native'
> 
> //
> <ScrollView 
>      style={{ 
>            width: "equal-to-or-less-than viewport width", 
>            height: "equal-to-or-less-than viewport height"
>      }}
>     contentContainerStyle={{"..."}}
>     horizontal
> >
>     <ScrollView 
>            style={{
>               width: "lower-than Parent-ScrollView width",
>               height: "lower-than Parent-ScrollView height"
>            }} 
>            contentContainerStyle={{"..."}}
>     >
>     </ScrollView>
> </ScrollVIew>
> ```
> 
> > NOTE: Always Set explicit boundaries(width/height) for Scrollview. So basically all sides of the border have to appear on the screen

This works perfectly, thanks for the explanation!

**Comment by Miriatowa** [comment]
I tried the following schemes: 
1.import { ScrollView } from 'react-native-gesture-handler' ，This solution can solve the sliding nesting problem, but some Android models, such as Xiaomi and Huawei, will report errors
2. This scheme will not solve the problem，like this：
`> - [ ] <scrollView vertical>
> - [ ] <scrollView horizontal >  <--- this scrollview component is not using just prevent scrollview same props error (so just add scrollEnabled to false ) thats it no need to add     nestedScrollEnabled
> - [ ] <scrollView vertical>
> - [ ] 
> - [ ] </scrollView vertical>
> - [ ] </scrollView horizontal>
> - [ ] </scrollView vertical>` 
3、So the way I'm doing it so far is when I click on the inner scrollview, I disable the outer scrollview, and when I click on the outer ScrollView, I enable the outer scrollView. Note that if scrollview does not allow scrolling, you need to use setNativeprops to directly change the underlying view, not setState. Asynchronous rendering would make it difficult to respond to state changes in a timely manner, resulting in delays and insensitivity

**Comment by Glaysonvisgueira** [comment]
> I tried the following schemes: 1.**import { ScrollView } from 'react-native-gesture-handler'** ，This solution can solve the sliding nesting problem, but some Android models, such as Xiaomi and Huawei, will report errors 2. **This scheme will not solve the problem**，like this： `<scrollView vertical> <scrollView horizontal > <scrollView vertical> </scrollView vertical> </scrollView horizontal> </scrollView vertical>` 3、So the way I'm doing it so far is when I click on the inner scrollview, **I disable the outer scrollview, and when I click on the outer ScrollView, I enable the outer scrollView**. Note that if scrollview does not allow scrolling, you need to use setNativeprops to directly change the underlying view, not setState. Asynchronous rendering would make it difficult to respond to state changes in a timely manner, resulting in delays and insensitivity

Have you tried putting sizes on all ScrollViews as explained above? For me it works fine.

**Comment by Miriatowa** [comment]
I tried that, but it didn't work for me

**Comment by Miriatowa** [comment]
> > I tried the following schemes: 1.**import { ScrollView } from 'react-native-gesture-handler'** ，This solution can solve the sliding nesting problem, but some Android models, such as Xiaomi and Huawei, will report errors 2. **This scheme will not solve the problem**，like this： `<scrollView vertical> <scrollView horizontal > <scrollView vertical> </scrollView vertical> </scrollView horizontal> </scrollView vertical>` 3、So the way I'm doing it so far is when I click on the inner scrollview, **I disable the outer scrollview, and when I click on the outer ScrollView, I enable the outer scrollView**. Note that if scrollview does not allow scrolling, you need to use setNativeprops to directly change the underlying view, not setState. Asynchronous rendering would make it difficult to respond to state changes in a timely manner, resulting in delays and insensitivity
> 
> Have you tried putting sizes on all ScrollViews as explained above? For me it works fine.

I tried that, but it didn't work for me

**Comment by Glaysonvisgueira** [comment]
> > > I tried the following schemes: 1.**import { ScrollView } from 'react-native-gesture-handler'** ，This solution can solve the sliding nesting problem, but some Android models, such as Xiaomi and Huawei, will report errors 2. **This scheme will not solve the problem**，like this： `<scrollView vertical> <scrollView horizontal > <scrollView vertical> </scrollView vertical> </scrollView horizontal> </scrollView vertical>` 3、So the way I'm doing it so far is when I click on the inner scrollview, **I disable the outer scrollview, and when I click on the outer ScrollView, I enable the outer scrollView**. Note that if scrollview does not allow scrolling, you need to use setNativeprops to directly change the underlying view, not setState. Asynchronous rendering would make it difficult to respond to state changes in a timely manner, resulting in delays and insensitivity
> > 
> > 
> > Have you tried putting sizes on all ScrollViews as explained above? For me it works fine.
> 
> I tried that, but it didn't work for me

Can you show your code?

**Comment by jessebwr** [comment]
We've finally are addressing this internally: https://github.com/facebook/react-native/pull/39097. Hopefully should be merged soon.

**Comment by zach-shute-cb** [comment]
Looks like this was merged. Any idea when it will go out or which version of RN will get it? I looked at the (CHANGELOG.md)[https://github.com/facebook/react-native/blob/main/CHANGELOG.md] and was not able to find any indication that this fix had gone live

**Comment by zach-shute-cb** [comment]
cc @jessebwr

**Comment by nonjee888** [comment]
I just upgrade RN version to 0.73.1 and checked, but still horizontal doesn't work for the nested  ScrollView in my code. Anyone has found it works horizontally?

**Comment by Nantris** [comment]
This still doesn't seem to work?
